Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -253,3 +253,179 @@ theorem Language.IsContextFree.reverse (L : Language T) :
   fun ⟨g, hg⟩ => ⟨g.reverse, hg ▸ Set.ext g.mem_reverse_language_iff_reverse_mem_language⟩
 
 end closure_reversal
+
+section embed_project
+
+/-! This section contains only auxiliary constructions that will shorten upcoming proofs of
+closure properties. When combining several grammars together, we usually want to take a sum type of
+their nonterminal types and embed respective nonterminals to this sum type.
+We subsequently show that the resulting grammar preserves derivations of those strings that may
+contain any terminal symbols but only the proper nonterminal symbols.
+The embedding operation must be injective.
+The projection operation must be injective on those symbols where it is defined. -/
+
+/-- Mapping `Symbol` when it is a nonterminal. -/
+def Symbol.map {N₀ N : Type*} (f : N₀ → N) : Symbol T N₀ → Symbol T N
+  | Symbol.terminal t => Symbol.terminal t
+  | Symbol.nonterminal n => Symbol.nonterminal (f n)
+
+/-- Mapping `Symbol` when it is a nonterminal; may return `none`. -/
+def Symbol.filterMap {N₀ N : Type*} (f : N → Option N₀) : Symbol T N → Option (Symbol T N₀)
+  | Symbol.terminal t => some (Symbol.terminal t)
+  | Symbol.nonterminal n => Option.map Symbol.nonterminal (f n)
+
+/-- Map the type of nonterminal symbols of a `ContextFreeRule` . -/
+def ContextFreeRule.map {N₀ N : Type*} (r : ContextFreeRule T N₀) (f : N₀ → N) :
+    ContextFreeRule T N :=
+  ⟨f r.input, r.output.map (Symbol.map f)⟩
+
+/-- A pair of `ContextFreeGrammar`s with maps between their types of nonterminal symbols that work,
+roughly speaking, in a good way. -/
+structure EmbeddedContextFreeGrammar (T : Type uT) where
+  /-- The smaller grammar. -/
+  g₀ : ContextFreeGrammar.{uN} T
+  /-- The bigger grammar. -/
+  g : ContextFreeGrammar.{uN} T
+  /-- Mapping nonterminals from the smaller type to the bigger type. -/
+  embedNT : g₀.NT → g.NT
+  /-- Mapping nonterminals from the bigger type to the smaller type. -/
+  projectNT : g.NT → Option g₀.NT
+  /-- The embedding map is injective. -/
+  embed_inj : Function.Injective embedNT
+  /-- The projecting map is injective where defined. -/
+  project_inj : ∀ x y : g.NT, projectNT x = projectNT y → x = y ∨ projectNT x = none
+  /-- The two mappings are essentially inverses. -/
+  projectNT_embedNT : ∀ n₀ : g₀.NT, projectNT (embedNT n₀) = some n₀
+  /-- Each rule of the smaller grammar has a corresponding rule in the bigger grammar. -/
+  embed_mem_rules : ∀ r : ContextFreeRule T g₀.NT, r ∈ g₀.rules → r.map embedNT ∈ g.rules
+  /-- Each rule of the bigger grammar whose input nonterminal is recognized by the smaller grammar
+  has a corresponding rule in the smaller grammar. -/
+  preimage_of_rules :
+    ∀ r : ContextFreeRule T g.NT,
+      r ∈ g.rules → ∀ n₀ : g₀.NT,
+        embedNT n₀ = r.input → ∃ r₀ ∈ g₀.rules, r₀.map embedNT = r
```

Comment:
Do you have a reference for this?

Reply:
I have a reference for a similar structure.

Reply:
https://arxiv.org/pdf/2302.06420 page 6